### PR TITLE
Create Blog “a-note-on-composing-components-with-react”

### PR DESCRIPTION
**Pull request by mfrachet**
Automatically generated by Netlify CMS

### CODE REVIEW

File: blog/a-note-on-composing-components-with-react.md [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+
+## What are we doing wrong?
+
+So here's the problem. For now, we are simply modifying the component we built in a vertical way: by adding a lots of new props on the component, and making a lot of conditional comparisons to determine the look of the card.
+
+There's another problem. Material Design Cards are not defined specifically. I mean, it's not supposed to provide a way to create a `PrimaryCard`, `TinyCard` and so forth. With the Material Design Card specs, you're supposed to be able to create a lots of different types of Cards without that much pain.
+
+What we have done until now actually breaks a part of the [Opened Close principle](https://en.wikipedia.org/wiki/Open%E2%80%93closed_principle). This principle explains that a code should be closed to modifications and opened to extensions (aka inheritance or composition).
+
+The idea is to avoid creating one single "thing" with a lot of complexe things inside, and to separate the concerns by creating multiple files.
```
[the remark this answers is not in the document]
Comment (ncuillery-youitv) [reply blog/a-note-on-composing-components-with-react.md]: maybe "entities" instead of "files"

File: blog/a-note-on-composing-components-with-react.md [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+## Material design cards using composition
+
+We have now a bit more of experience dealing with React component and we know that what we are doing is not sufficiently good for now. It's okay, and we'll see how we can improve using the composable nature of React.
```
Comment (ncuillery-youitv) [inline blog/a-note-on-composing-components-with-react.md]: We have **more** experience

Comment (ncuillery-youitv) [reply blog/a-note-on-composing-components-with-react.md]: with React component**s**

Comment (ncuillery-youitv) [reply blog/a-note-on-composing-components-with-react.md]: we know ~that~ what we are doing

Comment (ncuillery-youitv) [reply blog/a-note-on-composing-components-with-react.md]: "improve the usability"
"improve our Card component"
?

(there must be something after "improve")

File: blog/a-note-on-composing-components-with-react.md [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+First, let's point that a piece of UI doesn't necessarily means one component. It's not always a one for one matching. I invite you reading [Brad Frost Atomic Design](http://bradfrost.com/blog/post/atomic-web-design/) if you want more information on of to build complex UIs in a more abstract way.
```
Comment (ncuillery-youitv) [inline blog/a-note-on-composing-components-with-react.md]: Enough "let's" 😄 

~let's point~ 👉 It's worth noticing

Comment (ncuillery-youitv) [reply blog/a-note-on-composing-components-with-react.md]: I invite you **to read**

Comment (ncuillery-youitv) [reply blog/a-note-on-composing-components-with-react.md]: "on of to build" 🤔 

👉 "about building"

File: blog/a-note-on-composing-components-with-react.md [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+_The design we're going to implement is probably the best one but it gives a good idea of the impact of composition in React._
```
Comment (ncuillery-youitv) [inline blog/a-note-on-composing-components-with-react.md]: "is**n't** probably the best one" I guess

Comment (ncuillery-youitv) [reply blog/a-note-on-composing-components-with-react.md]: impact 👉 benefit?

File: blog/a-note-on-composing-components-with-react.md [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+
+Let's try to split the Card component in a composable manner so that we don't have to modify its implementation when we want to create new piece of UI:
```
Comment (ncuillery-youitv) [inline blog/a-note-on-composing-components-with-react.md]: No "Let's" 😉

Comment (ncuillery-youitv) [reply blog/a-note-on-composing-components-with-react.md]: so ~that~ we ~d~**w**on't have to modify its implementation ~when~**if** we want to create new piece of UI:

File: blog/a-note-on-composing-components-with-react.md [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+    <IconButton name="favorite" onClick={handleFav} />
+    <IconButton name="star" onClick={handleStar} />
+  </CardActions>
+</Card>
+```
+
+Using this approach, we're able to build any type of cards:
+
+
+
```
Comment (ncuillery-youitv) [inline blog/a-note-on-composing-components-with-react.md]: Why the triple carriage return from this point to the end of the post?

File: blog/a-note-on-composing-components-with-react.md [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+![Multiple material cards](/assets/cards003.png "Multiple material cards")
+
+
+
+## The smell I rely on
+
+So this is a way to rely on composability to avoid getting complicated and unmaintainable components in React.
```
Comment (ncuillery-youitv) [inline blog/a-note-on-composing-components-with-react.md]: ~rely on~ 👉 legerage

Comment (ncuillery-youitv) [reply blog/a-note-on-composing-components-with-react.md]: avoid ~getting~ complicated and unmaintainable components in React

File: blog/a-note-on-composing-components-with-react.md [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+
+
+
+In the future, if you have some components that start to grow, here are my "tricks" or at least the code smells (it's opinionated and depends on the cases) that indicates that I have to split my component on something a bit more composable:
```
Comment (ncuillery-youitv) [inline blog/a-note-on-composing-components-with-react.md]: on the **use-**cases

Comment (ncuillery-youitv) [reply blog/a-note-on-composing-components-with-react.md]: that indicate~s that~

File: blog/a-note-on-composing-components-with-react.md [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+
+* \- If a component has more than 5-6 props
+* \- If a component is more than 100 LoC
```
Comment (ncuillery-youitv) [inline blog/a-note-on-composing-components-with-react.md]: is 👉 contains

Comment (ncuillery-youitv) [reply blog/a-note-on-composing-components-with-react.md]: LoC 👉 lines of code

(the acronym may not be clear to anybody)

File: blog/a-note-on-composing-components-with-react.md [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+* \- Can I extract this component in isolation to play in another app?
```
Comment (ncuillery-youitv) [inline blog/a-note-on-composing-components-with-react.md]: Can I extract this component **and reuse it** in another app?

File: blog/a-note-on-composing-components-with-react.md [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+* \- Have I to tweak (or can I tweak) the props of the underlying components? (See this issue for more details [rn-placeholder#38](https://github.com/mfrachet/rn-placeholder/issues/38))
```
Comment (ncuillery-youitv) [inline blog/a-note-on-composing-components-with-react.md]: **Do I have**

File: blog/a-note-on-composing-components-with-react.md [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+* \- Is it easy to move an underlying component somewhere else?
```
Comment (ncuillery-youitv) [inline blog/a-note-on-composing-components-with-react.md]: I would rephrase this like the previous bullet and locate it right below:

>Can I extract this component and reuse it somewhere else?
>Can I extract any underlying component and reuse it somewhere else?

File: blog/a-note-on-composing-components-with-react.md [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+
+* \- If a component has more than 5-6 props
+* \- If a component is more than 100 LoC
+* \- Can I extract this component in isolation to play in another app?
+* \- Have I to tweak (or can I tweak) the props of the underlying components? (See this issue for more details [rn-placeholder#38](https://github.com/mfrachet/rn-placeholder/issues/38))
+* \- Is it easy to move an underlying component somewhere else? 
+
+
+
+But be careful! You have to find a balance between something fully composable but also easy to use. Remember that abstracting is not always a good thing.
```
Comment (ncuillery-youitv) [inline blog/a-note-on-composing-components-with-react.md]: **the right** balance

Comment (ncuillery-youitv) [reply blog/a-note-on-composing-components-with-react.md]: tell why

Maybe by ending the sentence with "if it makes your code more verbose and less readable"

File: blog/a-note-on-composing-components-with-react.md [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+path: composing-components
+date: 2019-05-16T18:47:58.055Z
+title: A note on composing components with React
+tags:
+  - react
+  - javascript
+---
+We often hear about composition when talking about frontend frameworks. Composition is the ability to create complex UIs by assembling component in as less uncoupled manner as possible.
+
+[Ryan Florence has made a video some years ago](https://www.youtube.com/watch?v=hEGg-3pIHlE) concerning compound components and the benefits of having well defined and composable APIs.
```
[the remark this answers is not in the document]
Comment (mfrachet) [reply blog/a-note-on-composing-components-with-react.md]: I'll check this 😊

File: blog/a-note-on-composing-components-with-react.md [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+  // new implementation
+  footerTitle="Footer title"
+  footerSchedules={['5pm', '7am', '2pm']}
+  footerAction={() => alert('Do footer stuff')}
+/>
+```
+
+So as you've probably mentioned, each time we want our card to have a different look, we modify its implementation. Now let's imagine that you are the owner and maintainer of such a library. Each time people want to have a different organisation on the Card, you are supposed to modify your component, publish it to npm and manage the versions.
+
+It's a lot of work.
```
[the remark this answers is not in the document]
Comment (mfrachet) [reply blog/a-note-on-composing-components-with-react.md]: Exactly, it's mentioned below 😊

File: blog/a-note-on-composing-components-with-react.md [hunk 16]
```diff
@@ -0,0 +1,125 @@
...
+
+## What are we doing wrong?
+
+So here's the problem. For now, we are simply modifying the component we built in a vertical way: by adding a lots of new props on the component, and making a lot of conditional comparisons to determine the look of the card.
+
+There's another problem. Material Design Cards are not defined specifically. I mean, it's not supposed to provide a way to create a `PrimaryCard`, `TinyCard` and so forth. With the Material Design Card specs, you're supposed to be able to create a lots of different types of Cards without that much pain.
```
[the remark this answers is not in the document]
Comment (mfrachet) [reply blog/a-note-on-composing-components-with-react.md]: I find this concept really hard to explain. The spec explains how we can create Cards, I mean different cards. There are no definition of a "standard" card. It's just a way to define some element that compose together.

File: blog/a-note-on-composing-components-with-react.md (hunk 6 above)
Comment (mfrachet) [reply blog/a-note-on-composing-components-with-react.md]: CMS 😅

File: blog/a-note-on-composing-components-with-react.md [hunk 17]
```diff
@@ -0,0 +1,142 @@
...
+  actionOne={() => console.log('Do job')}
+  actionSecond={() => console.log('Do another job')}
+  actionLike={() => console.log('Do action like')}
+  actionLove={() => console.log('Do action love')}
+/>
+```
+
+And to be honest, the previous code snippet is a valid code and it can work in real world applications.
+
+The problem comes when you start customising the Card to handle its variations. Because as I mentioned above, Card in the Material Design specification is a way to represent heterogeneous information using some homogeneous styles.
```
Comment (frinyvonnick) [inline blog/a-note-on-composing-components-with-react.md]:
```suggestion
The problem comes when you start customizing the Card to handle its variations. Because as I mentioned above, Card in the Material Design specification is a way to represent heterogeneous information using some homogeneous styles.
```

Comment (frinyvonnick) [inline blog/a-note-on-composing-components-with-react.md]: I feel like "As I mentionned" is enough you don't need "Because" 😄

File: blog/a-note-on-composing-components-with-react.md [hunk 18]
```diff
@@ -0,0 +1,142 @@
+---
+path: composing-components
+date: 2019-05-16T18:47:58.055Z
+title: A note on composing components with React
+tags:
+  - react
+  - javascript
+---
+We often hear about composition when talking about frontend frameworks. Composition is the ability to create complex UIs by assembling components in the less uncoupled manner as possible.
```
Comment (frinyvonnick) [inline blog/a-note-on-composing-components-with-react.md]: I don't understand this `in the less uncoupled manner as possible`. You mean `less coupled manner`, right ? Otherwise I would be glad to have further explanations 👍

File: blog/a-note-on-composing-components-with-react.md [hunk 19]
```diff
@@ -0,0 +1,142 @@
...
+
+[Ryan Florence has made a video some years ago](https://www.youtube.com/watch?v=hEGg-3pIHlE) concerning compound components and the benefits of having well-defined and composable APIs.
+
+
+
+<iframe width="100%" height="400" src="https://www.youtube.com/embed/hEGg-3pIHlE" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
+
+
+
+I'm a software developer and I make mistakes. And some years ago, I've released [rn-placeholder](https://github.com/mfrachet/rn-placeholder) using an API that was not extensible enough because of lack of composability.
```
Comment (frinyvonnick) [inline blog/a-note-on-composing-components-with-react.md]:
```suggestion
I'm a software developer and I make mistakes. Some years ago, I've released [rn-placeholder](https://github.com/mfrachet/rn-placeholder) using an API that was not extensible enough because of lack of composability.
```

And is unnecessary imho 😄

File: blog/a-note-on-composing-components-with-react.md [hunk 20]
```diff
@@ -0,0 +1,142 @@
...
+  //actionLike={() => console.log('Do action like')}
+  //actionLove={() => console.log('Do action love')}
+  // new implementation
+  footerTitle="Footer title"
+  footerSchedules={['5pm', '7am', '2pm']}
+  footerAction={() => console.log('Do footer stuff')}
+/>
+```
+
+So as you've probably noticed, each time we want our card to have a different look, we modify its implementation. Now imagine yourself as an owner or maintainer of such a library. Each time people want to have a different organisation on the Card, you are supposed to modify your component, publish it to npm and manage the versions.
```
Comment (frinyvonnick) [inline blog/a-note-on-composing-components-with-react.md]:
```suggestion
As you've probably noticed, each time we want our card to have a different look, we modify its implementation. Now imagine yourself as an owner or maintainer of such a library. Each time people want to have a different organisation on the Card, you are supposed to modify your component, publish it to npm and manage the versions.
```

So is more a spoken word, I guess ?

_Edit: You use it a few times later (in case you want to remove it 😉)_

File: blog/a-note-on-composing-components-with-react.md [hunk 21]
```diff
@@ -0,0 +1,142 @@
...
+
+It's a lot of work.
+
+## What are we doing wrong?
+
+So here's the problem. For now, we are simply modifying the component we built in a vertical way: by adding a lot of new props on the component, and making a lot of conditional comparisons to determine the look and feel of the card.
```
Comment (frinyvonnick) [inline blog/a-note-on-composing-components-with-react.md]:
```suggestion
So here's the problem. For now, we are simply modifying the component we built in a vertical way by adding a lot of new props on the component and making a lot of conditional comparisons to determine the look and feel of the card.
```

File: blog/a-note-on-composing-components-with-react.md [hunk 22]
```diff
@@ -0,0 +1,142 @@
...
+
+There's another problem. Material Design Cards are not defined specifically. I mean, it's not supposed to provide a way to create a `PrimaryCard`, `TinyCard` and so forth. With the Material Design Card specs, you're supposed to be able to create a lots of different types of Cards without that much pain.
+
+What we have done until now actually breaks a part of the [Opened Closed principle](https://en.wikipedia.org/wiki/Open%E2%80%93closed_principle). This principle tells that a piece of code should be closed to modifications and opened to extensions (aka inheritance or composition).
+
+The idea is to avoid creating one single "thing" with a lot of complex things inside, but separate the concerns by creating multiple entities.
+
+## Material design cards using composition
+
+We have more of experience dealing with React components and we know what we are doing is not sufficiently good for now. It's okay, and we'll see how we can improve the Card component API using the composable nature of React.
```
Comment (frinyvonnick) [inline blog/a-note-on-composing-components-with-react.md]:
```suggestion
We have more of experience dealing with React components and we know what we are doing is not good enough yet. It's okay, we'll see how we can improve the Card component API using the composable nature of React.
```

Comment (frinyvonnick) [reply blog/a-note-on-composing-components-with-react.md]: I found a [nice article](https://www.grammarly.com/blog/comma-before-and/) about comma before "and" 😄

File: blog/a-note-on-composing-components-with-react.md [hunk 23]
```diff
@@ -0,0 +1,142 @@
...
+
+First, it's worth noticing that a piece of UI doesn't necessarily means one component. It's not always a one for one matching. I invite you to read [Brad Frost Atomic Design](http://bradfrost.com/blog/post/atomic-web-design/) if you want more information about building complex UIs in a more abstract way.
```
Comment (frinyvonnick) [inline blog/a-note-on-composing-components-with-react.md]:
```suggestion
First, it's worth noticing that a piece of UI doesn't necessarily means one component. It's not always a one to one matching. I invite you to read [Brad Frost Atomic Design](http://bradfrost.com/blog/post/atomic-web-design/) if you want more informations about building complex UIs in a more abstract way.
```

File: blog/a-note-on-composing-components-with-react.md [hunk 24]
```diff
@@ -0,0 +1,142 @@
...
+
+_The design we're going to implement isn't probably the best one but it gives a good idea of the benefit of composition in React._
```
Comment (frinyvonnick) [inline blog/a-note-on-composing-components-with-react.md]:
```suggestion
_The design we're going to implement isn't probably the best one but it gives a good idea of the benefits of composition in React._
```

File: blog/a-note-on-composing-components-with-react.md [hunk 25]
```diff
@@ -0,0 +1,142 @@
...
+</Card>
+```
+
+Using this approach, we're able to build any type of cards:
+
+![Multiple material cards](/assets/cards003.png "Multiple material cards")
+
+## The smell I rely on
+
+So this is a way to leverage composability to avoid complicated and unmaintainable components in React.
```
Comment (frinyvonnick) [inline blog/a-note-on-composing-components-with-react.md]:
```suggestion
So this is a way to leverage composability to avoid complex and unmaintainable components in React.
```

File: blog/a-note-on-composing-components-with-react.md [hunk 26]
```diff
@@ -0,0 +1,142 @@
...
+
+In the future, if you have some components that start to grow, here are my "tricks" or at least the code smells (it's opinionated and depends on the use-cases) that indicates I have to split my component on something a bit more composable:
+
+* If a component has more than 5-6 props
+* If a component contains more than 100 lines of code
+* Can I extract this component and reuse it in another app?
+* Do I have to tweak (or can I tweak) the props of the underlying components? (See this issue for more details [rn-placeholder#38](https://github.com/mfrachet/rn-placeholder/issues/38))
+* Can I move the underlying component somewhere else in the parent component (top or bottom)? 
+
+But be careful! You have to find the right balance between something fully composable but also easy to use. Remember that abstracting is not always a good thing.
```
Comment (frinyvonnick) [inline blog/a-note-on-composing-components-with-react.md]:
```suggestion
But be careful! You have to find the right balance between something fully composable and easy to use. Remember that abstracting is not always a good thing.
```

File: blog/a-note-on-composing-components-with-react.md [hunk 27]
```diff
@@ -0,0 +1,142 @@
...
+
+
+
+If you find yourself writing tons of line to create a component you know you'll use only one time, it's probably not necessary to find a good abstraction. However if it's a component that will leave in multiple places and that can probably move in terms of shape, you maybe have to think twice while implementing.
```
Comment (frinyvonnick) [inline blog/a-note-on-composing-components-with-react.md]:
```suggestion
If you find yourself writing tons of lines to create a component you know you'll use it only one time, it's probably not necessary to find a good abstraction. However if it's a component that you will use in multiple places and that can probably move in terms of shape, you maybe have to think twice while implementing.
```

File: blog/a-note-on-composing-components-with-react.md [hunk 28]
```diff
@@ -0,0 +1,142 @@
...
+
+
+
+And don't forget that with the new composable API, we can create a component that just looks like the old one!
```
Comment (frinyvonnick) [inline blog/a-note-on-composing-components-with-react.md]:
```suggestion
And don't forget with the new composable API, we can create a component that just looks like the old one!
```

File: blog/a-note-on-composing-components-with-react.md (hunk 18 above)
Comment (mfrachet) [reply blog/a-note-on-composing-components-with-react.md]: Yep 😅 THanks for spotting this!
